kdc/httpkadmind.c

```diff
@@ -0,0 +1,2025 @@
+             const char *unparsed)
+{
+    krb5_error_code ret = 0;
+    krb5_keytab_entry key;
+    size_t i;
+
+    if (princ->n_key_data <= 0)
+        return 0;
+
+    memset(&key, 0, sizeof(key));
+    for (i = 0; ret == 0 && i < princ->n_key_data; i++) {
+        krb5_key_data *kd = &princ->key_data[i];
+
+        if (kadm5_all_keys_are_bogus(1, kd))
+            continue;
+
+        key.principal = princ->principal;
+        key.vno = kd->key_data_kvno;
+        key.keyblock.keytype = kd->key_data_type[0];
+        key.keyblock.keyvalue.length = kd->key_data_length[0];
+        key.keyblock.keyvalue.data = kd->key_data_contents[0];
+
+        /*
+         * XXX kadm5 doesn't give us set_time here, but we can compute it using
+         * the KeyRotation metadata in the TL data.  We should!
+         */
+        key.timestamp = time(NULL);
+
+        ret = krb5_kt_add_entry(r->context, r->keytab, &key);
+    }
+    if (ret)
+        krb5_warn(r->context, ret,
+                  "Failed to write keytab entries for %s", unparsed);
+
+    return ret;
+}
+
+static void
+random_password(krb5_context context, char *buf, size_t buflen)
+{
+    static const char chars[] =
+        "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789.,";
+    char p[32];
+    size_t i;
+    char b;
+
+    buflen--;
+    for (i = 0; i < buflen; i++) {
+        if (i % sizeof(p) == 0)
+            krb5_generate_random_block(p, sizeof(p));
+        b = p[i % sizeof(p)];
+        buf[i] = chars[b % (sizeof(chars) - 1)];
+    }
+    buf[i] = '\0';
+}
+
+static krb5_error_code
+make_kstuple(krb5_context context,
+             kadm5_principal_ent_rec *p,
+             krb5_key_salt_tuple **kstuple,
+             size_t *n_kstuple)
+{
+    size_t i;
+
+    *kstuple = 0;
+    *n_kstuple = 0;
+
+    if (p->n_key_data < 1)
+        return 0;
+    *kstuple = calloc(p->n_key_data, sizeof (*kstuple));
+    for (i = 0; *kstuple && i < p->n_key_data; i++) {
+        if (p->key_data[i].key_data_kvno == p->kvno) {
+            (*kstuple)[i].ks_enctype = p->key_data[i].key_data_type[0];
+            (*kstuple)[i].ks_salttype = p->key_data[i].key_data_type[1];
+            (*n_kstuple)++;
+        }
+    }
+    return *kstuple ? 0 :krb5_enomem(context);
+}
+
+/* Get keys for one principal */
+static krb5_error_code
+get_keys1(kadmin_request_desc r, const char *pname)
+{
+    kadm5_principal_ent_rec princ;
+    krb5_key_salt_tuple *kstuple = NULL;
+    krb5_error_code ret = 0;
+    krb5_principal p = NULL;
+    uint32_t mask =
+        KADM5_PRINCIPAL | KADM5_KVNO | KADM5_MAX_LIFE | KADM5_MAX_RLIFE |
+        KADM5_ATTRIBUTES | KADM5_KEY_DATA | KADM5_TL_DATA;
+    uint32_t create_mask = mask & ~(KADM5_KEY_DATA | KADM5_TL_DATA);
+    size_t nkstuple = 0;
+    int change = 0;
+    int refetch = 0;
+    int freeit = 0;
+
+    memset(&princ, 0, sizeof(princ));
+    princ.key_data = NULL;
+    princ.tl_data = NULL;
+
+    ret = krb5_parse_name(r->context, pname, &p);
```

Comment:
```suggestion
    ret = krb5_parse_name_flags(r->context, pname,
                                KRB5_PRINCIPAL_PARSE_NO_DEF_REALM,
                                &p);
```
Then if the resulting principal has a NULL realm, use the requested realm if requested.

Reply:
Leaving this for some other day.  But let's leave a breadcrumb about this.